### PR TITLE
Allow for LocalFile in task chain input data

### PR DESCRIPTION
**Pull request by mesmith75**
Allows for LHCbTransform to chain input data with LocalFile in case of batch running.

### CODE REVIEW

**Review by egede — COMMENTED** [review]
Why is this limited to spcific datafiles. Should it not be able to work with all GangaFile objects? MassStorageFile would be another sensible file type to use here.

File: ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py [hunk 1]
```diff
@@ -197,11 +199,30 @@ def createChainUnit(self, parent_units, use_copy_output=True):
                             for pat in excl_pat_list:
                                 if re.search(pat, f.lfn) and "LFN:" + f.lfn in flist:
                                     flist.remove("LFN:" + f.lfn)
+                    elif isType(f, LocalFile):
```
Comment (egede) [inline ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py]: Should istype be isinstance instead to allow for inheritance?

Comment (mesmith75) [reply ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py]: yes

**Comment by mesmith75** [comment]
I think you are right, we should allow for MassStorageFile as well. But I guess we still want to keep the inputdata to a single type?

**Comment by egede** [comment]
> I think you are right, we should allow for MassStorageFile as well. But I guess we still want to keep the inputdata to a single type?

Is this not part of a larger somewhat unresolved issue in Ganga where we allow inputdata to be a mix of different types of `GangaFile` objects. Often things will fail if that is the case. I do not think this is the place where we want to enforce that restriction (as it implies to a broader area).

File: ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py (hunk 1 above)
Comment (mesmith75) [reply ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py]: Actually no - these files are wrapped up in proxies so `isinstance` doesn't work on them

**Comment by mesmith75** [comment]
So this now takes LocalFile and MassStorageFile.

File: ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py [hunk 2]
```diff
@@ -197,11 +199,30 @@ def createChainUnit(self, parent_units, use_copy_output=True):
                             for pat in excl_pat_list:
                                 if re.search(pat, f.lfn) and "LFN:" + f.lfn in flist:
                                     flist.remove("LFN:" + f.lfn)
+                    elif isType(f, LocalFile) or isType(f, MassStorageFile):
```
Comment (egede) [inline ganga/GangaLHCb/Lib/Tasks/LHCbTransform.py:203]: MassStorageFile is not imported.

**Review by egede — CHANGES_REQUESTED** [review]

**Review by egede — APPROVED** [review]

**Pull request merged**